Feat/#445 implement ErrorOutOfGasCodeStore and ErrorMaxCodeSizeExceeded error state (#475)
privacy-scaling-explorations/zkevm-specs: specs/error_state/ErrorCodeStore.md

# ErrorCodeStore state

## Procedure
`ErrorCodeStore` is a combined error code and it's designed to handle code store related errors: `CodeStoreOutOfGas` and `MaxCodeSizeExceeded`. This type of error only occurs when executing `CREATE`/`CREATE2` opcode or a deployment transaction (tx.to = null).

### EVM behavior

> While handling a `CREATE`/`CREATE2` opcode, initial bytecode will be executed and then current call context is created. The contract bytecode will be returned through `RETURN` opcode as the execution result. `RETURN` returns a specified memory chunk [`offset`...`offset` + `length`] as bytecode. The gas cost for storing the bytecode is,

Reviewer: Nitpick: "is," -> "is:"

Author: fixed in [d630df6](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/475/commits/d630df687049ed065be459186951093609f74b15)

# ErrorCodeStore state

## Procedure

> `ErrorCodeStore` is for two original code store related errors: `CodeStoreOutOfGas` and `MaxCodeSizeExceeded`. This type of error only occurs when executing create(create,create2) op code or tx deploy transaction(tx.to = null).

Reviewer: I know this was not changed in this PR, but perhaps a bit of language cleaning could be done in this PR? Like "op code" -> "opcode" or perhaps deleting "tx" (I think it's redundant).

Author: fixed in [6ad3b04](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/475/commits/6ad3b041f054e6bed0b1a292bcaacc6e95cd6ecf)

Author: I was hesitanting to refine the sentence or not, but I agree with you, we should do language cleaning in this PR. Have a fix, [6ad3b04](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/475/commits/6ad3b041f054e6bed0b1a292bcaacc6e95cd6ecf), please help review it's more clear for you or not.

### EVM behavior

> When handling a CREATE-kind transaction, Initial bytecode opcodes will run and current call context is created. The final bytecode opcodes to store for new contract is the `RETURN` opcode of init codes result.

Reviewer: Similarly as above, it would be great if this text is a bit improved. It's difficult to understand (for example, what does it mean "initial byte code opcodes will run"?)

Author: fixed in [6ad3b04](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/475/commits/6ad3b041f054e6bed0b1a292bcaacc6e95cd6ecf)

> `RETURN` opcode returns memory [`offset`...`offset` + `length`] content as bytecode to store into state db. For returned bytecode, store them cost additional gas.   

Reviewer: Similarly as above, for example I think "them" is redundant.

Author: fixed in [6ad3b04](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/475/commits/6ad3b041f054e6bed0b1a292bcaacc6e95cd6ecf)

```
let CODE_DEPOSIT_BYTE_COST = 200
code_store_cost = CODE_DEPOSIT_BYTE_COST * len(bytecodes)
``` 

- If `code_store_cost` > gas left, it is `CodeStoreOutOfGas` case.
- If returned bytecode length > `MAXCODESIZE`, it is `MaxCodeSizeExceeded` case.  

> In circuit bus mapping side, check these two code store errors in [here](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/bus-mapping/src/circuit_input_builder/input_state_ref.rs#L1148&L1155). When executing opcode is `RETURN` and call context is creating(`call.is_create == true`) meanwhile.  

Reviewer: Not changed in this PR, but I am not sure what "meanwhile" refers to?

Author: I think it means `at the same time`

> Even though errors occur in `CREATE` kind opcodes, it is special not checking error in executing opcode `CREATE` directly. Circuit implementation takes similar strategy, not constrain error directly in CREATE opcodes, but in `RETURN` step context. Following this way it easy to get the key property state `length` and construct constraints against it.

Reviewer: "it is special not checking" -> "we do not check"?
"not constrain error" -> "it does not constrain error"?
"this way it easy" -> "this way it is easy"

Author: I just re-write it in [6ad3b04](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/475/commits/6ad3b041f054e6bed0b1a292bcaacc6e95cd6ecf). Please help review it make more sense or not.

# ErrorCodeStore state

## Procedure

> `ErrorCodeStore` is a combined error code and it's designed to handle code store related errors: `CodeStoreOutOfGas` and `MaxCodeSizeExceeded`. This type of error only occurs when executing `CREATE`/`CREATE2` opcode or a deployment transaction (tx.to = null).

Reviewer 2: ```suggestion
`ErrorCodeStore` is a combined error code for handling the CodeStoreOutOfGas` and `MaxCodeSizeExceeded` code store related errors. This type of error only occurs when executing `CREATE`/`CREATE2` opcode or a deployment transaction (tx.to = null).
```

### EVM behavior

> While handling a `CREATE`/`CREATE2` opcode, initial bytecode will be executed and then current call context is created. The contract bytecode will be returned through `RETURN` opcode as the execution result. `RETURN` returns a specified memory chunk [`offset`...`offset` + `length`] as bytecode. The gas cost for storing the bytecode is:

Reviewer 2: I find this paragraph a bit hard to read. Here is a suggestion (of my understanding), please correct me if I am wrong.

```
let CODE_DEPOSIT_BYTE_COST = 200
code_store_cost = CODE_DEPOSIT_BYTE_COST * len(bytecodes)
``` 

> - If `code_store_cost` > gas left, it is `CodeStoreOutOfGas` case.

Reviewer 2: ```suggestion
Which error are we handling?
- If `code_store_cost` > gas left, `ErrorCodeStore` corresponds to `CodeStoreOutOfGas`;
```

> - If returned bytecode length > `MAXCODESIZE`, it is `MaxCodeSizeExceeded` case.  

Reviewer 2: ```suggestion
- If returned bytecode length > `MAXCODESIZE`, it corresponds to `MaxCodeSizeExceeded`.
```

> On the circuit bus mapping side, the checks for these two code store errors are [here](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/8a633f7c3de2da72f0817def57c1703241cced97/bus-mapping/src/circuit_input_builder/input_state_ref.rs#L1296-L1304). This error happens only when current opcode is `RETURN` and it's a `CREATE`/`CREATE2` call (`call.is_create == true`). We can't get contract bytecode length in `CREATE`/`CREATE2` opcodes and it's only available in `RETURN` opcode so we handle these two errors in `RETURN` opcode.

Reviewer 2: ```suggestion
On the circuit bus mapping side, the checks for these two code store errors are [here](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/8a633f7c3de2da72f0817def57c1703241cced97/bus-mapping/src/circuit_input_builder/input_state_ref.rs#L1296-L1304). This error only happens when both the current opcode is `RETURN` and the call is either `CREATE` or `CREATE2` call, that is when `call.is_create == true`. 

Even though these are `CREATE`/`CREATE2` errors, as we cannot get the contract bytecode length when processing these opcodes, we handle them in the `RETURN` opcode where the bytecode length is available.
```

Reviewer 2: It might be worthwhile also adding a comment in the create.py file that the  the `CodeStoreOutOfGas` and `MaxCodeSizeExceeded`  errors are handled in the `RETURN` opcode instead.

Overall it looks like the following:  
- Pop EVM word `offset` and `length` from the stack, 
- Go to `ErrorCodeStore` state when call context is being created & select which one of the followings occurs:
  1. Storing `length` of bytecode runs out of gas.
  2. `length` of bytecode exceeds `MAXCODESIZE`.

### Constraints
1. Current opcode is `RETURN` and `is_create` is `True`.
2. `code_store_cost` > gas_left or `length` > `MAXCODESIZE`

> 3. Current call must be failed.

Reviewer 2: ```suggestion
3. Current call fails.
```

4. Current call is not static call.

> 5. If it's a root call, it transits to `EndTx`

Reviewer 2: ```suggestion
5. If it's a root call, it transits to `EndTx`. Otherwise the caller's context is restored by reading the `rw_table`, then it does step state transition to it.
```

Author: I prefer not to make this change in this PR bcs these two points (5 and 6) are more like a fixed pattern and all the error state MD files using the same description. If we would like to revise them, we should do it in another refinment PR to revise all. What do you think?

Reviewer 2: Sure!

> 6. if it is not root call, it restores caller's context by reading to `rw_table`, then does step state transition to it.

Reviewer 2: ```suggestion
```

Author: same as above

- If `code_store_cost` > gas left, it is `CodeStoreOutOfGas` case.
- If returned bytecode length > `MAXCODESIZE`, it is `MaxCodeSizeExceeded` case.  

> In circuit bus mapping side, check these two code store errors in [here](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/8a633f7c3de2da72f0817def57c1703241cced97/bus-mapping/src/circuit_input_builder/input_state_ref.rs#L1296-L1304). This error happens only when current opcode is `RETURN` and it's a `CREATE`/`CREATE2` call (`call.is_create == true`). We can't get contract bytecode length in `CREATE`/`CREATE2` opcodes and it's only available in `RETURN` opcode so we handle these two errors in `RETURN` opcode.

Reviewer: Nitpick: perhaps "On the circuit bus mapping side, the checks for these two code store errors are [here]"?

Author: fixed in [d630df6](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/475/commits/d630df687049ed065be459186951093609f74b15)
